# Массивы, списки и коллекции

Посмотрите полное [видео по C# 101](https://www.youtube.com/watch?v=qLeF_wpnVto&list=PLdo4fOcmZ0oVxKLQCHpiUWun7vlJJvUiN&index=12) для этого модуля.

Массивы, списки и коллекции могут быть весьма полезны. Рассмотрим список:
> Запустите следующий код. Распечатывает ли он то, что вы ожидали?

In [1]:
using System;
using System.Collections.Generic;

var names = new List<string> { "<name>", "Ana", "Felipe" };
foreach (var name in names)
{
    Console.WriteLine($"Hello {name.ToUpper()}!");
}

Hello <NAME>!
Hello ANA!
Hello FELIPE!


## Так что же означает этот код?

- **System.collections.Generic**: это пространство имен, в котором есть списки. Если вы не сообщите коду, что используете его, вам придется писать "Systems.Collections.Generic.List" каждый раз, когда вы хотите использовать список. Это экономит время на вводе текста!
- **var**: это то, что вы пишете, когда у вас есть переменная, но вы не знаете/не заботитесь о ее типе.
- **List\<string\>**: это означает, что вы создаете список строк. Вместо `string` вы можете указать `int`, `double` или любую другую переменную.
- **foreach**: это еще один цикл for! Он проходит через каждый элемент в списке.
- **name in names**: этот стиль предпочитают многие люди. `names` — это весь список, содержащий имена во множественном числе. `name` — это отдельный элемент в `names`.

## Альтернативный метод

Предыдущий код немного более понятен для чтения, чем код ниже, но код ниже имеет более узнаваемый код, основанный на том, что мы узнали.
На самом деле это просто два разных стиля написания одного и того же кода. Не стесняйтесь использовать то, что имеет для вас больший смысл!
> Запустите следующий код.
>
> Можете ли вы определить схожие части кода в двух разных методах?
>
> Какой метод вы предпочитаете?

In [3]:
using System;
using System.Collections.Generic;

var names = new List<string> { "<name>", "Ana", "Felipe" };
for (int i = 0;i < names.Count; i++)
{
    Console.WriteLine($"Hello {names[i].ToUpper()}");
}

Hello <NAME>
Hello ANA
Hello FELIPE


## Add

Вы можете легко добавлять имена в списки. В списках есть метод `Add()`, который добавляет новый элемент в конец списка.
> Запустите код.
>
> Затем попробуйте заменить имя в коде на свое собственное.

In [4]:
var names = new List<string> { "<name>", "Ana", "Felipe" };

names.Add("Sophia");

foreach (var name in names)
{
    Console.WriteLine($"Hello {name.ToUpper()}!");
}

Hello <NAME>!
Hello ANA!
Hello FELIPE!
Hello SOPHIA!


## Remove

Вы также можете удалять имена. Попробуйте:

> Запустите код.
>
> Затем выберите, какое имя вы хотите удалить.
>
> Что произойдет, если вы попытаетесь удалить то, чего нет?

In [6]:
var names = new List<string> { "<name>", "Ana", "Felipe" };

names.Remove("<name>");

for (int i = 0;i < names.Count; i++)
{
    Console.WriteLine($"Hello {names[i].ToUpper()}");
}

Hello ANA
Hello FELIPE


## Вывод определенного элемента

Что делать, если вы не хотите выводить именя всех своих друзей? Что, если вы просто хотите вывести имя одного друга? Вот тут-то и приходят на помощь скобки.
> Запустите код.
>
> Попробуйте вывести имя под другим номером в списке.
>
> Нужен ли вам 0 или 1, чтобы вывести первый элемент списка?

In [1]:
var names = new List<string> { "<name>", "Ana", "Felipe" };
Console.WriteLine($"Hello {names[0].ToUpper()}");

Hello <NAME>


## Поиск

В последней теме вы узнали, как найти, какой элемент хранится по определенному индексу. Теперь, зная предмет, узнаем его индекс.

> Запустите код
>
> На каком индексе Sophia?
>
> Какой индекс у "Scott"?

In [11]:
var names = new List<string> { "<name>", "Sophia", "Felipe", "Scott"};
Console.WriteLine(names[3]);

Scott


Не забывайте, что списки основаны на "0". Первое место является "0-м".

# Playground

Поиграйте с тем, что вы узнали! Вот несколько идей:
> Составьте список необходимых вам продуктов под названием `groceries`.
>
> Можете ли вы вывести "buy \<grocery\>!" для каждого предмета?
>
> Какой третий пункт списка? Это то же самое, что `groceries[3]?`
>
> Обязательно добавьте "toothpaste".
>
> Удалите наименее любимый предмет.
>
> А что третье в списке?

In [3]:
using System;
using System.Collections.Generic;
var groceries = new List<string> {"Potato", "Carrot", "tomato"};
groceries.Add("Toothpaste");
groceries.Remove("Toothpaste");
foreach (var i in groceries) 
{
    Console.WriteLine($"Buy {i}");
}
Console.WriteLine($"Третий элемент списка - {groceries[2]}");

Buy Potato
Buy Carrot
Buy tomato
Третий элемент списка - tomato


# Поиск, сортировка и индексирование списков

Посмотрите полное [видео по C# 101](https://www.youtube.com/watch?v=NJ5ghiutzfY&list=PLdo4fOcmZ0oVxKLQCHpiUWun7vlJJvUiN&index=13) для этого модуля.

In [1]:
using System;
using System.Collections.Generic;
var names = new List<string> { "Sophia", "Ana", "Jayme", "Bill" };
string name = "Ana";
var index = names.IndexOf(name);
Console.WriteLine($"Found {name} at {index}");

Found Ana at 1


## Что означает -1?

Если `IndexOf()` возвращает -1, это означает, что не удалось найти элемент в списке. Фактически, вы можете сделать небольшой оператор if, который не позволит найти элемент:

> Запустите код.
>
> Попробуйте несколько разных имен.

In [2]:
var names = new List<string> { "Sophia", "Ana", "Jayme", "Bill" };
string name = "Bill";
var index = names.IndexOf(name);
if(index == -1){
    Console.WriteLine($"{name} not found");
} else {
    Console.WriteLine($"Found {names[index]} at {index}");
}

Found Bill at 3


## Сортировка

До сих пор вы просто вводили имена в случайном порядке. Но иногда полезно отсортировать список. `Sort()` берет список и организует его. Он просматривает типы переменных и организует их наиболее разумным способом: если это строки, он сортирует их в алфавитном порядке, если это числа, то он организует их от меньшего к большему.

Обратите внимание, что вам не нужно писать `sortedList = name.Sort()`, вам просто нужно написать `names.Sort()`. `Sort()` изменяет сам список, и вам не нужно сохранять действие в новом объекте.

> Запустите код!
>
> Не стесняйтесь добавлять любые другие имена, чтобы их можно было отсортировать.

In [3]:
var names = new List<string> { "Sophia", "Ana", "Jayme", "Bill" };
Console.WriteLine("Pre Sorting:");
foreach(var name in names )
{
    Console.WriteLine(name);
}

names.Sort();

Console.WriteLine();
Console.WriteLine("Post Sorting:");
foreach(var name in names )
{
    Console.WriteLine(name);
}

Pre Sorting:
Sophia
Ana
Jayme
Bill

Post Sorting:
Ana
Bill
Jayme
Sophia


# Playground

Поиграйте с тем, что вы узнали! Вот несколько идей:

> Составьте список продуктов и рассортируйте их! какой индекс у 'Carrots'?
>
> Попробуйте составить список чисел! Сортируются ли они так, как вы ожидаете?
>
> Распечатайте списки, которые вы составили.

In [11]:
var names = new List<string> { "Sophia", "Ana", "Jayme", "Bill", "Carrots"};
names.Sort();
foreach (string el in names)
{
    Console.WriteLine(el);
}

Ana
Bill
Carrots
Jayme
Sophia


In [8]:
var numbers = new List<int> { 100, 5, 30, 4, 5};
numbers.Sort();
foreach (int el in numbers)
{
    Console.WriteLine(el);
}


4
5
5
30
100


# Списки других типов

Посмотрите полное [видео по C# 101](https://www.youtube.com/watch?v=oIQdb93xewE&list=PLdo4fOcmZ0oVxKLQCHpiUWun7vlJJvUiN&index=14) для этого модуля.

Вы практиковались в составлении списков строк, но составить список можно из чего угодно! Вот пример числового списка.

## Фибоначчи

Фибоначчи — это классная числовая последовательность. Он складывает последние два числа, чтобы получить следующее число. Вы начинаете с 1 и 1

1 + 1 = 2 (1, 1, 2)
1 + 2 = 3 (1, 1, 2, 3)
2 + 3 = 5 (1, 1, 2, 3, 5)
3 + 5 = 8 (1, 1, 2, 3, 5, 8)
и так далее. В природе существует множество вещей, которые соответствуют этой числовой последовательност!

> Начните с базовых чисел: вот список, в котором всего 1, 1. Запустите его и посмотрите, что произойдет.

In [23]:
var fibonacciNumbers = new List<long> { 1, 1 };
for (int i = 0; i < 20; i ++)
{
    fibonacciNumbers.Add(fibonacciNumbers[i] + fibonacciNumbers[i + 1]);
    Console.WriteLine(fibonacciNumbers[i]);
}
    

1
1
2
3
5
8
13
21
34
55
89
144
233
377
610
987
1597
2584
4181
6765


Теперь вам не нужны только 1,1! Вы хотите больше последовательности. В этом коде вы используете два последних числа списка, складываете их вместе, чтобы получить следующее число, а затем добавляете его в список.

> Запустите код, чтобы опробовать его.

In [16]:
var fibonacciNumbers = new List<int> {1, 1}; // Starting the list off with the basics

var previous = fibonacciNumbers[fibonacciNumbers.Count - 1]; // Take the last number in the list
var previous2 = fibonacciNumbers[fibonacciNumbers.Count - 2]; // Take the second to last number in the list

fibonacciNumbers.Add(previous + previous2); // Add the previous numbers together, and attach the sum to the end of the list

foreach (var item in fibonacciNumbers) // Print out the list
    Console.WriteLine(item);

1
1
2


## Count -1

Почему вам нужно выполнить `fibonacciNumbers.Count -1`, чтобы получить последний номер списка? Ну, `Count` сообщает вам, сколько элементов находится в списке. Однако индекс элемента начинается с нуля. Итак, если бы в вашем списке был только один элемент, счетчик был бы один, но индекс элемента был бы 0. Индекс и счетчик последнего элемента всегда равны единице.

# Задача: Фибоначчи до 20-го числа

Мы предоставили вам базу кода, связанную с Фибоначчи. Можете ли вы составить список из первых 20 чисел Фибоначчи?

> Составьте и выведите список, содержащий первые 20 чисел Фибоначчи.

In [24]:
var fibonacciNumbers = new List<long> { 1, 1 };
for (int i = 0; i < 20; i ++)
{
    fibonacciNumbers.Add(fibonacciNumbers[i] + fibonacciNumbers[i + 1]);
    Console.WriteLine(fibonacciNumbers[i]);
}

1
1
2
3
5
8
13
21
34
55
89
144
233
377
610
987
1597
2584
4181
6765


## Секреты и уловки

- Итоговое число должно быть 6765.
- Сможете ли вы сделать цикл `for`? Цикл `foreach`? Цикл `while`? Какой тип цикла вы предпочитаете и какой будет более полезным?
- Вы уже близки, но отстаете на одну цифру? Это действительно распространенная проблема! Помните, что `>` и `>=` похожи, но в конечном итоге они отличаются друг от друга. Попробуете поиграть с этим?
- Помните, что вы уже начинаете с двумя элементами в списке.
- В тупике? Посмотрите [видео C# 101](https://www.youtube.com/watch?v=oIQdb93xewE&list=PLdo4fOcmZ0oVxKLQCHpiUWun7vlJJvUiN&index=14) для этого модуля. Попробуйте сделать паузу, как только у вас появится идея, и сначала опробовать ее, прежде чем продолжать просмотр.